diff --git a/profiling-2/VX_csr_data.sv b/profiling-2/VX_csr_data.sv
index b1e6843..f40968a 100644
--- a/profiling-2/VX_csr_data.sv
+++ b/profiling-2/VX_csr_data.sv
@@ -190,6 +190,10 @@ import VX_fpu_pkg::*;
                         `VX_CSR_MPM_SCHED_ID_H      : read_data_ro_r = 32'(pipeline_perf_if.sched_idles[`PERF_CTR_BITS-1:32]);
                         `VX_CSR_MPM_SCHED_ST        : read_data_ro_r = pipeline_perf_if.sched_stalls[31:0];
                         `VX_CSR_MPM_SCHED_ST_H      : read_data_ro_r = 32'(pipeline_perf_if.sched_stalls[`PERF_CTR_BITS-1:32]);
+                        `VX_CSR_MPM_SCHED_AW        : read_data_ro_r = pipeline_perf_if.sched_active_warp_count[31:0];
+                        `VX_CSR_MPM_SCHED_AW_H      : read_data_ro_r = 32'(pipeline_perf_if.sched_active_warp_count[`PERF_CTR_BITS-1:32]);
+                        `VX_CSR_MPM_SCHED_SW        : read_data_ro_r = pipeline_perf_if.sched_stalled_warp_count[31:0];
+                        `VX_CSR_MPM_SCHED_SW_H      : read_data_ro_r = 32'(pipeline_perf_if.sched_stalled_warp_count[`PERF_CTR_BITS-1:32]);
                         `VX_CSR_MPM_IBUF_ST         : read_data_ro_r = pipeline_perf_if.ibf_stalls[31:0];
                         `VX_CSR_MPM_IBUF_ST_H       : read_data_ro_r = 32'(pipeline_perf_if.ibf_stalls[`PERF_CTR_BITS-1:32]);
                         `VX_CSR_MPM_SCRB_ST         : read_data_ro_r = pipeline_perf_if.scb_stalls[31:0];
